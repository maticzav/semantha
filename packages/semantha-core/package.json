{
  "name": "semantha-core",
  "version": "0.0.0-semantha",
  "main": "dist/index.js",
  "repository": "https://github.com/maticzav/semantha.git",
  "scripts": {
    "clean": "rimraf dist",
    "dev": "ts-node src/index.ts",
    "test": "NODE_ENV=test jest",
    "build": "tsc -d",
    "prepublish": "npm-run-all build"
  },
  "dependencies": {
    "@octokit/rest": "^16.1.0",
    "semver": "^5.6.0"
  },
  "devDependencies": {
    "@types/jest": "23.3.13",
    "@types/node": "10.17.50",
    "jest": "23.6.0",
    "npm-run-all": "4.1.5",
    "ts-jest": "23.10.5",
    "ts-node": "7.0.1",
    "tslint": "5.12.1",
    "tslint-config-prettier": "1.17.0",
    "tslint-config-standard": "8.0.1",
    "typescript": "3.2.4"
  },
  "license": "MIT"
}
