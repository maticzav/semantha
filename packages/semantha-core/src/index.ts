export { analyzeCommits, SemanthaRelease, SemanthaRule } from './analyser'
export { constants, SemanthaVersion } from './constants'
export {
  getCommitsSinceLastRelease,
  GithubFile,
  GithubCommit,
  GithubRepository,
} from './github'
export { publishWorkspace } from './publish'
export { loadWorkspace, Workspace } from './workspaces'
