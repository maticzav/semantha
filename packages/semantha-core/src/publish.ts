/**
 *
 * Publishes workspace with new version.
 *
 * @param path
 */
export async function publishWorkspace(path: string): Promise<any> {
  return
}
