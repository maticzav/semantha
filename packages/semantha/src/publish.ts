import execa from 'execa'
import * as fs from 'fs'
import _ from 'lodash'
import * as path from 'path'
import {
  SemanthaRelease,
  Dependency,
  getNextVersion,
  Package,
} from 'semantha-core'
import url from 'url'
import { filterMap } from './utils'

/**
 *
 * Write package.json to package.
 *
 * @param pkgPath
 * @param pkg
 */
export function writePackage(
  pkgPath: string,
  pkg: string,
): { status: 'ok' } | { status: 'err'; message: string } {
  const pjPath = path.resolve(pkgPath, 'package.json')
  try {
    fs.writeFileSync(pjPath, pkg)
    return { status: 'ok' }
  } catch (err) {
    return { status: 'err', message: err.message }
  }
}

export interface PublishOptions {
  registry: string
  token: string
}

/**
 *
 * Publishes release to NPM.
 *
 * @param release
 */
export async function publish(
  release: SemanthaRelease,
  releases: SemanthaRelease[],
  options: PublishOptions,
): Promise<
  | { status: 'ok'; release: SemanthaRelease }
  | { status: 'err'; message: string }
> {
  try {
    /* Update package.json and override the old one */

    const pj = applyVersionsToPackage(release.workspace.pkg, releases)
    const updated = writePackage(release.workspace.path, pj)

    if (updated.status === 'err') {
      return {
        status: 'err',
        message: `Error writing package.json: ${updated.message}`,
      }
    }

    /* Authenticate */

    const authentication = auth(
      release.workspace.path,
      options.registry,
      options.token,
    )

    if (authentication.status === 'err') {
      return {
        status: 'err',
        message: `Error authenticating: ${authentication.message}`,
      }
    }

    /* Publish */

    execa('npm', ['publish', '--registry', options.registry], {
      cwd: release.workspace.path,
    })

    return { status: 'ok', release: release }
  } catch (err) {
    return { status: 'err', message: err.message }
  }

  /* Helper functions */

  /**
   * Taken from npm/cli.
   *
   * Maps a URL to an identifier.
   *
   * Name courtesy schiffertronix media LLC, a New Jersey corporation
   *
   * @param {String} uri The URL to be nerfed.
   * @returns {String} A nerfed URL.
   */
  function toNerfDart(uri: string): string {
    var parsed = url.parse(uri)
    delete parsed.protocol
    delete parsed.auth
    delete parsed.query
    delete parsed.search
    delete parsed.hash

    return url.resolve(url.format(parsed), '.')
  }

  /**
   *
   * Composes a npmrc file with auth.
   *
   * @param pkgPath
   * @param registry
   * @param token
   */
  function auth(
    pkgPath: string,
    registry: string,
    token: string,
  ): { status: 'ok' } | { status: 'err'; message: string } {
    const npmrcPath = path.resolve(pkgPath, '.npmrc')
    try {
      const npmrc = `${toNerfDart(registry)}:_authToken = ${token}`
      fs.writeFileSync(npmrcPath, npmrc)

      return { status: 'ok' }
    } catch (err) {
      return { status: 'err', message: err.message }
    }
  }
}

/**
 *
 * Updates package.json file of a released package.
 *
 * @param release
 */
export function applyVersionsToPackage(
  pkg: Package,
  releases: SemanthaRelease[],
): string {
  /* Updates the package version */

  const pkgRelease = releases.find(
    release => release.workspace.pkg.name === pkg.name,
  )!

  const pjWithVersion = updateVersion(pkg.raw, getNextVersion(pkgRelease)!)

  /**
   * Compares package dependencies with releases and updates matches
   * to the latest calculated version.
   */
  const dependencies = filterMap<Dependency, Dependency>(dependency => {
    const release = releases.find(
      release => release.workspace.pkg.name === dependency.name,
    )

    /* Generates a new dependency if found */
    if (release) {
      return {
        name: dependency.name,
        type: dependency.type,
        version: getNextVersion(release)!,
      }
    } else {
      return null
    }
  }, pkg.dependencies)

  const pjWithDependencies = dependencies.reduce((acc, dependency) => {
    return updateDependency(acc, dependency)
  }, pjWithVersion)

  return pjWithDependencies

  /* Helper functions */
  /**
   * Thanks Rhys Arkins for explaining how renovate works.
   * All functions below are inspired by how renovate works.
   */

  /**
   *
   * Updates package version in package.json file.
   *
   * @param raw
   * @param dependency
   */
  function updateVersion(raw: string, version: string): string {
    /**
     * First update version in parsed package.json, then
     * repeat updating raw package.json files until parsed versions
     * match.
     */
    let parsed = JSON.parse(raw)
    const oldVersion = parsed.version

    /* Update parsed package.json */
    parsed.version = version

    const searchString = `"${oldVersion}"`
    const newString = `"${version}"`

    let newContent = null
    let searchIndex = raw.indexOf(`"version"`) + 'version'.length

    for (; searchIndex < raw.length; searchIndex += 1) {
      if (matchAt(raw, searchIndex, searchString)) {
        const testContent = replaceAt(raw, searchIndex, searchString, newString)

        if (_.isEqual(parsed, JSON.parse(testContent))) {
          newContent = testContent
          break
        }
      }
    }

    return newContent!
  }

  /**
   *
   * Updates dependency version in package.json file.
   *
   * @param raw
   * @param dependency
   */
  function updateDependency(raw: string, dependency: Dependency): string {
    /**
     * First update dependencies in parsed package.json, then
     * repeat updating raw package.json files until parsed versions
     * match.
     */
    let parsed = JSON.parse(raw)
    const oldVersion = parsed[dependency.type][dependency.name]

    /* Update parsed package.json */
    parsed[dependency.type][dependency.name] = dependency.version

    const searchString = `"${oldVersion}"`
    const newString = `"${dependency.version}"`

    let newContent = null
    let searchIndex =
      raw.indexOf(`"${dependency.type}"`) + dependency.type.length

    for (; searchIndex < raw.length; searchIndex += 1) {
      if (matchAt(raw, searchIndex, searchString)) {
        const testContent = replaceAt(raw, searchIndex, searchString, newString)

        if (_.isEqual(parsed, JSON.parse(testContent))) {
          newContent = testContent
          break
        }
      }
    }

    return newContent!
  }

  /**
   *
   * @param content
   * @param index
   * @param match
   */
  function matchAt(content: string, index: number, match: string): boolean {
    return content.substring(index, index + match.length) === match
  }

  /**
   *
   *
   *
   * @param content
   * @param index
   * @param oldString
   * @param newString
   */
  function replaceAt(
    content: string,
    index: number,
    oldString: string,
    newString: string,
  ): string {
    return (
      content.substr(0, index) +
      newString +
      content.substr(index + oldString.length)
    )
  }
}
