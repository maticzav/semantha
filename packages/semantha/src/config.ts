import globby from 'globby'
import parseGithubUrl from 'parse-github-url'
import { sync as readPkg } from 'read-pkg'
import { GithubRepository, Workspace } from 'semantha-core'
import { withDefault } from './utils'

export interface Configuration {
  repository: GithubRepository
  workspaces: Workspace[]
}

/**
 *
 * Finds Semantha configuration in a particular cwd.
 *
 * @param cwd
 */
export async function getConfigurationFrom(
  cwd: string,
): Promise<
  { status: 'ok'; config: Configuration } | { status: 'err'; message: string }
> {
  try {
    /** Find configuration package.json file */
    const pkg = readPkg({ cwd: cwd, normalize: true })

    if (!pkg.repository || !pkg.workspaces) {
      return {
        status: 'err',
        message: 'Missing workspaces or repository definition in package.json',
      }
    }

    /** Determines repository */

    const repository = getRepositoryFromURL(pkg.repository.url)

    if (repository.status === 'err') {
      return {
        status: 'err',
        message: "Something's wrong with your repository definition.",
      }
    }

    /* Finds workspaces */

    const workspaces = await findAllWorkspacesForPaths(pkg.workspaces, cwd)

    if (workspaces.status === 'err') {
      return { status: 'err', message: workspaces.message }
    }

    return {
      status: 'ok',
      config: {
        repository: repository.repo,
        workspaces: workspaces.workspaces,
      },
    }
  } catch (err) {
    return { status: 'err', message: err.message }
  }

  /* Helper functions */

  /**
   *
   * Obtains repository information from repository URL.
   *
   * @param url
   */
  function getRepositoryFromURL(
    url: string,
  ): { status: 'ok'; repo: GithubRepository } | { status: 'err' } {
    const repository = parseGithubUrl(url)

    if (repository && repository.owner && repository.name) {
      return {
        status: 'ok',
        repo: { owner: repository.owner, repo: repository.name },
      }
    } else {
      return { status: 'err' }
    }
  }

  /**
   *
   * Finds definitions for workspaces in specified cwd
   *
   * @param patterns
   * @param cwd
   */
  async function findAllWorkspacesForPaths(
    patterns: string[],
    cwd: string,
  ): Promise<
    | { status: 'ok'; workspaces: Workspace[] }
    | { status: 'err'; message: string }
  > {
    try {
      /** Find all paths */
      const paths = await globby(patterns, {
        cwd: cwd,
        onlyDirectories: true,
        gitignore: true,
        absolute: true,
      })

      /** Hydrate workspaces */
      const workspaces = await Promise.all(paths.map(hydrateWorkspace))

      return { status: 'ok', workspaces: workspaces }
    } catch (err) {
      return { status: 'err', message: err.message }
    }
  }

  /**
   *
   * Hydrates workspace in a given path by obtaining its
   * package.json definition.
   *
   * @param path
   */
  async function hydrateWorkspace(path: string): Promise<Workspace> {
    const pkg = await readPkg({ cwd: path })
    return {
      path,
      pkg: {
        name: pkg.name,
        version: pkg.version,
        dependencies: withDefault({})(pkg.dependencies),
        devDependencies: withDefault({})(pkg.devDependencies),
      },
    }
  }
}
