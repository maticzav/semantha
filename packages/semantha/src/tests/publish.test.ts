import * as fs from 'fs'
import * as path from 'path'
import { SemanthaRelease } from 'semantha-core'

import { loadPackage } from '../config'
import { applyVersionsToPackage, writePackage } from '../publish'

describe('write package', () => {
  test('reports error', async () => {
    const pkgPath = path.resolve(__dirname, './__fixtures__/packages/random/')
    const res = writePackage(pkgPath, 'hey')

    const pjPath = path.resolve(pkgPath, 'package.json')

    expect(res).toEqual({
      status: 'err',
      message: `ENOENT: no such file or directory, open '${pjPath}'`,
    })
  })

  test('correctly updates package.json', async () => {
    const pkgPath = path.resolve(__dirname, './__fixtures__/packages/write/')
    const res = writePackage(pkgPath, '{ "name": "package" }')

    expect(res).toEqual({
      status: 'ok',
    })
    expect(
      fs.readFileSync(path.resolve(pkgPath, 'package.json'), 'utf-8'),
    ).toBe('{ "name": "package" }')
  })
})

describe('applyVersionToPackage', () => {
  test('correctly updates package.json', async () => {
    const pkgPath = path.resolve(__dirname, './__fixtures__/packages/valid/')
    const pkg = await loadPackage(pkgPath)

    if (pkg.status === 'err') {
      fail()
      return
    }

    const releases: SemanthaRelease[] = [
      {
        impactingCommits: [],
        releaseType: { type: 'major' },
        workspace: {
          path: '/path',
          pkg: {
            raw: '',
            name: 'package',
            version: '1.0.0',
            dependencies: [],
          },
        },
      },
      {
        impactingCommits: [],
        releaseType: { type: 'major' },
        workspace: {
          path: '/path',
          pkg: {
            raw: '',
            name: 'test-dependency-a',
            version: '1.0.0',
            dependencies: [],
          },
        },
      },
      {
        impactingCommits: [],
        releaseType: { type: 'ignore' },
        workspace: {
          path: '/path',
          pkg: {
            raw: '',
            name: 'test-dependency-b',
            version: '1.0.0',
            dependencies: [],
          },
        },
      },
      {
        impactingCommits: [],
        releaseType: { type: 'prerelease', tag: 'alpha' },
        workspace: {
          path: '/path',
          pkg: {
            raw: '',
            name: 'test-dependency-c',
            version: '1.0.0',
            dependencies: [],
          },
        },
      },
    ]

    const updated = applyVersionsToPackage(pkg.pkg, releases)

    expect(updated).toMatchSnapshot()
  })
})
