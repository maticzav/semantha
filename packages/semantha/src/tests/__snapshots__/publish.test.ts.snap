// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publish applyVersionToPackage correctly updates package.json 1`] = `
"{
  \\"devDependencies\\": {
    \\"test-dependency-b\\": \\"1.0.0\\",
    \\"test-dependency-c\\": \\"1.0.1-alpha.0\\"
  },
  \\"name\\": \\"package\\",
  \\"scripts\\": {},
  \\"version\\": \\"2.0.0\\",
  \\"dependencies\\": {
    \\"test-dependency-a\\": \\"2.0.0\\",
    \\"irrelevant\\": \\"2.0.0\\"
  }
}
"
`;
